### ISSUE TITLE
Support server-sent events

### ISSUE THREAD

**Issue by EvanHahn** [fastify/light-my-request#279]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

I would like this module to support [server-sent events (SSE)](https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface).

(If helpful, I can submit a pull request.)

### Motivation

Currently, the best way to test SSE doesn't use this package at all. It seems the state-of-the-art is to start the server and use an `EventSource`, possibly with the [eventsource package](https://www.npmjs.com/package/eventsource). This works but requires the server to be active, which this module tries to avoid.

It would be nice if this package had an answer for testing SSE.

This was briefly discussed in #105.

### Example

There are several possible ways this API could be implemented. My favorite idea is a re-implementation of the W3C `EventSource`; something like this:

```javascript
const source = new inject.EventSource(dispatch, { path: '/my-events' })

source.onmessage = (event) => {
  console.log(event)

  if (event.data === 'please close') {
    source.close()
  }
}
```

But there are many possible options, which we can discuss (if we even want to implement this feature at all!).

**Comment by mcollina** [comment]
This would be amazing. I don't think WHAWG EventSource fits well with Node.js, I think a node `stream.Readable` would be better. 

Something like:

```js
const res = await inject(dispatch, { path: '/my-events', sse: true })

for await (const chunk of res) { ... }
```

Or analog.

**Comment by Uzlopak** [comment]
I dont think this is the right approach. 

I considered yesterday after reading this issue to implement a sse plugin for fastify.  And when I thought about it, inject method would be specific to the implementation of the plugin and not specific to light-my-request.

I implemented a fastify implementation of smee.io https://github.com/probot/smee.io/pull/155

A concept would be implementing a channel concept. So you keep the open SSE connections in a Map, where key is the channel and values would be an array of connections. So you register a route to open the SSE/EventSource connection. Basically an route handler, where you return a fastify.sseCreateEventSourceStream('channelId'), where the fastify Instance was decorated with a sseCreateEventSourceStream-function which returns a stream. So when you call that path, you instantiate a connection and start to consume it. You are free to define the channelId as you like: take it from a path parameter, a header or whatever. 

Then you want to send events. For this I would implement a fastify.sseEmitEvent('channelId', type, data), where the fastify instance was decorated with a sseEmitEvent function, which sends to a specific channel specific events with the provided data. 

This makes it possible to send data into the EventSourceStream from other route handlers. Or like in the case of smee.io on POST /:channelId sending the serialized body to the channelId. 

Then you would only need to use a pub sub method, like in my smee.io implementation, to share the payload between multiple fastify instances. 

If it would be implemented like I propose it, it would not need any changes in light-my-request. And emitting sseEvents could be done via the sseEmitEvent method.

Just my 2 cents.

**Comment by EvanHahn** [comment]
> I considered yesterday after reading this issue to implement a sse plugin for fastify. And when I thought about it, inject method would be specific to the implementation of the plugin and not specific to light-my-request.

@Uzlopak That seems like it'd work if this module were being used exclusively with Fastify, but this module can (theoretically) be used with any Node web server. I still think it's useful for this module to support SSE in some way.

I'm new to the Fastify world, so I may be totally wrong here.

**Comment by EvanHahn** [comment]
Some thoughts about the API design for this module (not Fastify overall):

- SSE doesn't quite work like a normal request:
  - You can't set `method` (and therefore can't set `payload`)
  - The response won't necessarily have a full payload
  - `signal` doesn't really make sense because you can close the `EventSource` at any time
  - You might wish to set some SSE-specific options, such as the `Last-Event-ID` or reconnection settings
- Users should be able to close the connection at any time.
- Users should be able to add event listeners for `message` and `error`. They should also probably be able to [listen to custom events](https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events-intro).
- Users should be able to detect why a connection was closed, if it was closed by the server. This might happen immediately but might happen after receiving many messages.

I think this merits a new method, rather than adding an option to `inject`. I also think this shouldn't use a typical `Response` object, because it doesn't really make sense.

I'd start by creating `inject.sse`, which returns an SSE source object:

```javascript
// inject.sse returns synchronously.
// It takes most of the same options as `inject`, but some are removed.
const source = inject.sse(dispatch, { path: '/my-events' })
```

You can listen to various events:

```javascript
source.on('error', (err) => {
  console.warn('An error occurred', err)
})
source.on('message', ({ data, id, event, retry }) => {
  console.log(`Message ID ${id}: event`)
})
source.on('custom-message', ({ data, id, event, retry }) => {
  console.log('Got a custom message')
})
source.on('open', () => {
  console.log('Event source opened')
})
source.on('close', (reason) => {
  console.log(`Event source was closed for reason ${reason}`)
})
```

At any time, you can close the source:

```javascript
source.close()
```

You can also use Node's built-in `on` to achieve an iterable style:

```javascript
import { on } from 'node:events';

// Get all messages until a stopping point.
// You can also listen to custom messages the same way.
const messages = []
for await (const message of on(source, 'message')) {
  messages.push(message)
  if (message.data === 'please stop') break
}
```

When you're all done, you must close the source. (I don't love this and would love it to be done automatically; perhaps there's a way to do this with `FinalizationRegistry`.)

```javascript
source.close()
```

What do you think?

**Comment by Uzlopak** [comment]
Ok, granted that light-my-source is framework independent. But i have trouble to understand why a mock eventsource server should be part of this module.

**Comment by EvanHahn** [comment]
Apologies for any confusion. I don't think a mock SSE server should be part of this module, but I think a mock SSE client/consumer should be.

**Comment by Uzlopak** [comment]
I really dont understand the use case. 

The reason of light my request is to inject a request to the instance and then the handler returns a response. 

If we want to be close to the real world, it means we make a request against a path (inject) and we should get a response which is basically a eventsource server. 

You say, that this is not what you need. You basically want a mocked eventsource client, where you can emit any event you want.

Lets say we implement a mock client, which should be actually really easy, because I implemented a real client  in undici last week and basically can be used as a blue print and with small modifications it should work as a mock.

Now. What real world case would we actually make testable by having a mock client?

It doesnt even touch any logic of the underlying web server. And the only thing i could imagine would be testing a browser eventsource client, which is not even the target of this package.

**Comment by EvanHahn** [comment]
I want to test SSE endpoints without starting the server.

Currently, our code [starts the web server on a local port](https://github.com/digidem/mapeo-map-server/blob/5371f3e507a5e7510f76871089402f63af7b77a1/test/e2e/imports.test.js#L80), [creates an `EventSource` object](https://github.com/digidem/mapeo-map-server/blob/5371f3e507a5e7510f76871089402f63af7b77a1/test/test-helpers/import-sse.js) to gather events, and then stops the server. This means we have special workaround code for SSE.

Ideally, we'd be able to run our SSE tests just like we run our others: without starting a web server.

Again, apologies if I'm not understanding something here.

**Comment by hamptonsmith** [comment]
Hey folks! A useful starting place might just be to provide some guidance on how we're meant to use `payloadAsStream`? My initial experimentation seems to show that inject() promises don't resolve (or callback not called) until after the HTTP handler returns, but in the case of SSE we naturally want to see the contents of the stream while the handler is running, then be able to terminate the faked connection from the client side at the end of the test. If there is a way to gain access to the Response stream before the handler resolves, then we could plug it into, e.g., [eventsource](https://www.npmjs.com/package/eventsource) so that this module doesn't need its own mock

**Comment by hamptonsmith** [comment]
After a bit of fiddling, my confusion seems to arise because the Response is not returned until the first byte is written. I.e., consider:

```javascript
import http from 'http';
import inject from 'light-my-request';

const dispatch = async function (req, res) {
    setInterval(() => res.write('data: ' + Date.now()), 5000);
};

(async function main() {
    setInterval(() => console.log('tick'), 1000);

    const { stream } = await inject(dispatch, {
        method: 'get',
        url: '/' ,
        payloadAsStream: true
    });

    console.log('Got stream');

    stream().on('data', d => console.log('data', d.toString('utf8')));
})().catch(console.error);
```

The output here is:

```
tick
tick
tick
tick
Got stream
data data: 1754150602011
tick
```

Note that sending a header is enough, but in a very default setup with [Fastify](https://www.npmjs.com/package/fastify) + [fastify-sse-v2](https://www.npmjs.com/package/fastify-sse-v2), no headers are sent until the first event is raised and `light-my-request` doesn't return a request.

Is this intended behavior? It creates a deadlock situation during my tests, since I'm waiting for a stream to trigger an event, but can't get the stream until after the first event is triggered.

**Comment by Uzlopak** [comment]
Wow. That plugin is interesting, but also a little bit missing the point.

Also about one and a half years later since my last comment on this thread the request makes also more sense to me. 

SSE/EventSource is unidirectional. At a former employer we used socket.io. And we used the socket connection also unidirectional... (for whatever reason). So it was basically just a fancy EventSource... We also used the "room"-feature of socket.io to send the information out. We used mongodb and basically all resources where having organisationId and a projectId, and users had a userId. Those were ObjectIds. As they were ensured to be unique, I used them as roomIds. We spawned mutiple api server instances, which were also socket.io-servers. The api servers were connected to rabbitMQ and processed the specific eventQueue. 

So if a user logged in and connected to the socket.io-server, based on the granted access, started to listen to various rooms, atleast the organizationId room and his personal userId room. 
So what happened was that rest api calls were sent to and processed by the system. We decided that some events are organization-Wide, project-Wide or user-specific, and dependending on set the specific Id it as roomId and emitted the event into the eventqueue in rabbitMQ. The api/socket servers - as I said - were subscribed to the specific eventqueue and on arrival of the message checked if somebody was in the messageheaders (rabbitMQ specific thing)  specified room(id), and if so, deserialized the message and sent it to the room. If not, flush and ack the message, and listen for the next message. It was really fast and stable.


But what I want to say, is that, I miss an approach regarding organizing somehow the data in that plugin. Sending Data with SSE is really not a biggy. It is only 20 lines of code. If you have a solution on how to distribute the data to the specific clients, then I would say, that it would be useful. 
Also you need to consider to use of some EventSource mechanics. E.g. `lastEventId`. You should e.g. consider to not just increment it. Nope. E.g. your events should have a unique ID, which does not need to be incremental. Now you connect via EventSource and the resource which the user is viewing in the browser is up to date, we just get messages if somebody else is changing something with the resource, or maybe if I do it in a different browser tab. Can we just update the data in the redux store or whatever via the provided (partial) data, or is the stored information somehow totally obsolete, so that we should make another rest api call and update the information regarding that resource in the browser?  Maybe the etag header is somewhat related to the lastEventId or the payload, so that you set the etag header to the request and maybe avoid loading data, which you already have?

Sorry, I  am babbling...

I wrote long time ago a SSE implementation in low level nodejs for smee.io and also as I already mentioned implemented the EventSource client in nodejs. So i am kind of knee deep into the topic.

So basically you open a connection and you "stream". I guess there is an misunderstanding here. stream in the sense of returning a stream object and stream in sense of continous sending of data.

I have to check the codebase, but it confuses me that you await the inject. I think the inject resolves when the request is ended by doing `res.end()` in nodejs. But what happens if you set returnAsStream as true... will it resolve anyway and you get the continously data as payload? Seems a little bit strange.

Can you provide me a repository so that i can play around, please?

**Comment by mcollina** [comment]
> Is this intended behavior? It creates a deadlock situation during my tests, since I'm waiting for a stream to trigger an event, but can't get the stream until after the first event is triggered.

Looks like a bug in that module.

**Comment by hamptonsmith** [comment]
But you're using `server.listen()` and `request()`, bypassing light-my-request injection: https://github.com/platformatic/mcp/blob/1e489fe96ba0d290de93059ea807f0f7b510cdab/test/last-event-id.test.ts#L184C49-L184C56

My example [here](https://github.com/fastify/light-my-request/issues/279#issuecomment-3146588130) doesn't use fastify or fastify-sse-v2.

**Comment by mcollina** [comment]
Here is another one: https://github.com/platformatic/mcp/blob/1e489fe96ba0d290de93059ea807f0f7b510cdab/test/redis-integration.test.ts#L168-L208.

**Comment by mcollina** [comment]
> My example https://github.com/fastify/light-my-request/issues/279#issuecomment-3146588130 doesn't use fastify or fastify-sse-v2.

This works correct for me, I'm seeing the actual output.

note that if you want to get a response immediately before the first write you have to add:

```js
    res.writeHead(200, {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive'
    })
```

(which fastify-sse-v2 seems to lack)

**Comment by hamptonsmith** [comment]
Aha, yes. Those headers are sent [here in fastify-sse-v2](https://github.com/mpetrunic/fastify-sse-v2/blob/master/src/plugin.ts#L25-L27), but they are not sent until the first event is raised, which I agree is weird and likely non-compliant. I [filed a bug against `fastify-sse-v2`](https://github.com/mpetrunic/fastify-sse-v2/issues/97) to see if we can get it addressed.

A little more clarity in the docs of exactly how `payloadAsStream` changes the semantics of when `inject()` returns would be awesome (e.g., `curl -N` is able to stream me the events from https://github.com/fastify/light-my-request/issues/279#issuecomment-3146588130 just fine), but I agree we're now talking about "testing servers that violate HTTP semantics" and that's pretty out in the weeds. ;) Regardless, thanks for the help debugging!
